librbd/cache/pwl: code cleanup and some minor fixes

CongMinYin opened the pull request:
These are some minor modifications:

- Merge `rwl` initialization error handling and add `pmemobj_close(m_log_pool)`.
- Update comment.
- Update documentation.
- Clean up bool `waited_lanes`, `waited_entries`, `waited_buffers`.


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

CongMinYin commented:
@idryomov @ideepika @trociny Hi, this pr include some minor modifications, such as update log, comment, documentation. Could you please review it. CC @majianpeng .

CongMinYin commented:
jenkins test make check

ideepika commented:
> @idryomov @ideepika @trociny Hi, this pr include some minor modifications, such as update log, comment, documentation. Could you please review it. CC @majianpeng .

Thanks @CongMinYin added to my todo

ideepika commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -339,14 +335,12 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
                  << " expected " << MIN_WRITE_ALLOC_SIZE << dendl;
-      on_finish->complete(-EINVAL);
-      return false;
+      goto error_handle;
```
@CongMinYin what do you think about encapsulating this in a function like `handle_close`  https://github.com/ceph/ceph-ci/blob/92d7bc2869a92c18eda336e8aad551a77d26d08c/src/librbd/ImageState.cc#L876-L883

CongMinYin replied:
The two situations are a little different. The example is to process all the results in the callback function, and the code is concise. Here is only one kind of result processing. There are four kinds of results, and just what I changed now need to call `pmemobj_close(m_log_pool);`, the sucessful one even not in this func:
```
on_finish->complete(-errno);
on_finish->complete(-pmemobj_tx_errno());
on_finish->complete(-EINVAL);
m_image_ctx.op_work_queue->queue(on_finish, 0);  // sucess, not in this func
```
If I just change this kind of error handle to func:

```
void handle_error(PMEMobjpool *m_log_pool, Context *on_finish) {
    pmemobj_close(m_log_pool);
    on_finish->complete(-EINVAL);
}
initialize_pool() {
    ...
   if (...) {
        handle_error(m_log_pool, on_finish);
        return false;
    }
}
```
It doesn't seem to be getting simpler. If you don't like `goto` statements, may leave it as it is and add `pmemobj_close(m_log_pool) `before each `onfinish`. TODO: Make the same changes to SSD mode.

CongMinYin commented on src/librbd/cache/pwl/ssd/WriteLog.cc:
```diff
@@ -446,8 +446,8 @@ void WriteLog<I>::append_scheduled_ops(void) {
-  bool ops_remain = false; //no-op variable for SSD
-  bool appending = false; //no-op variable for SSD
+  bool ops_remain = false;  //unused, no-op variable for SSD
+  bool appending = false;   //unused, no-op variable for SSD
```
- updated this comment. 
- Add `pmemobj_close(m_log_pool)` after `on_finish->complete(-pmemobj_tx_errno())`, forgot it.
- Add a new commit to clean up `waited_lanes`, `waited_entries`, `waited_buffers`.

Please review again, thanks.

trociny commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -475,7 +475,7 @@ bool WriteLog<I>::retire_entries(const unsigned long int frees_per_tx) {
+    auto entry = m_log_entries.empty() ? nullptr : m_log_entries.front();
     while (!m_log_entries.empty() &&
            retiring_entries.size() < frees_per_tx &&
            this->can_retire_entry(entry)) {
```
nit: For me a little cleaner approach seems to use `this->can_retire_entry(m_log_entries.front())` here, and move `auto entry = m_log_entries.front();` inside this `while` block as the first statement (and remove `entry = m_log_entries.front();` at the end of the block). What do you think?

CongMinYin replied:
Good suggestion, thanks.  At the same time, it prevents getting the empty `m_log_entries.front() ` at the end. 
I have an idea to explore further. if `auto entry ` is in the `while` block, it should allocate for registers and release them every loop. Would it be better to put the declaration outside the `while` block. For example, I declare on `while` without assignment. Is this would be better and be allowed?
```
std::shared_ptr<GenericLogEntry> entry;
while {}
```

trociny replied:
Personally, I don't like optimization like this unless it gives a real gain, and I have no idea (or rather doubt) if it does in this case. Actually, I would expect some optimization from the compiler here that would make programmer not bother about things like this.

And it would make the code look a bit confusing: after this change, when looking at the code my first thought would be that the `entry` was going to be used outside the loop and I would spend some time looking at the code and trying to understand where it was.

Anyway, I don't have a strong opinion about this and would be fine with any final version.

CongMinYin replied:
Agree. The compiler may have optimized, and this makes the code harder to read. Then keep it that way.

trociny commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -319,6 +318,7 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
       this->m_free_log_entries = 0;
       lderr(cct) << "failed to initialize pool (" << this->m_log_pool_name << ")" << dendl;
```
nit: wouldn't it be useful to see `pmemobj_tx_errno()` in this error message?

trociny commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -319,6 +318,7 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
       lderr(cct) << "failed to initialize pool (" << this->m_log_pool_name << ")" << dendl;
       on_finish->complete(-pmemobj_tx_errno());
+      pmemobj_close(m_log_pool);
       return false;
```
nit: I would introduce `int r = -EINVAL` variable and store `-pmemobj_tx_errno()` there, use `on_finish->complete(r)` in `error_handle`, so  could use `goto error_handle` here too to have it more uniform.

trociny commented on src/librbd/cache/pwl/AbstractWriteLog.cc:
```diff
@@ -525,7 +525,8 @@ void AbstractWriteLog<I>::pwl_init(Context *on_finish, DeferredContexts &later)
              (access(m_log_pool_name.c_str(), F_OK) != 0)) {
-    ldout(cct, 5) << "Can't find the existed pool file " << m_log_pool_name << dendl;
+    lderr(cct) << "Can't find the existed pool file "
+                  << m_log_pool_name << dendl;
```
nit: would be nice to add `cpp_strerror(-errno)` to the error message.

trociny reviewed (approved):
Otherwise LGTM.

CongMinYin commented:
jenkins retest this please

CongMinYin commented:
Updated according to comments.

trociny commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -269,7 +267,7 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
                         this->m_log_pool_size,
                         (S_IWUSR | S_IRUSR))) == NULL) {
-      lderr(cct) << "failed to create pool (" << this->m_log_pool_name << ")"
+      lderr(cct) << "Failed to create pool (" << this->m_log_pool_name << ")"
                  << pmemobj_errormsg() << dendl;
```
I don't know if it was not in the initial version, or I just did not notice this. Anyway, the librbd convention seems to be alway use lowercase in debug/error messages. Why did you decided to change it here in other messages to start with uppercase?

CongMinYin replied:
Sorry, I noticed that some places use uppercase and some places use lowercase. I unified them into uppercase. Now it has been unified into lowercase. Thanks, I learned.

trociny commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -317,9 +314,10 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
-      on_finish->complete(-pmemobj_tx_errno());
-      return false;
+      lderr(cct) << "Failed to initialize pool: " << this->m_log_pool_name
+                 << ". Pmemobj TX errno: " << pmemobj_tx_errno() << dendl;
```
Just noticed after the second review, that there is `pmemobj_errormsg()`. So just wonder, could it be used here or if not may be there is something similar for  `pmemobj_tx_errno()`? If not, please just ignore. It looks good enough.

CongMinYin replied:
I don't know if transaction errors will be the same with pmemobj_errormsg(). But I don't think this is very important. Later, it is planned to change `libpmemobj` to `libpmem`. This PR I had @ you. And the transaction will be removed. Now let's keep it as it is.

CongMinYin commented:
jenkins test docs

ideepika commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -270,8 +267,8 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
-      lderr(cct) << "failed to create pool (" << this->m_log_pool_name << ")"
-                 << pmemobj_errormsg() << dendl;
+      lderr(cct) << "failed to create pool: " << this->m_log_pool_name
+                 << ". error:" << pmemobj_errormsg() << dendl;
```
missing space aftter error, << ". error: "

ideepika commented on doc/rbd/rbd-operations.rst:
```diff
@@ -10,7 +10,7 @@
    Live-Migration <rbd-live-migration>
    Persistent Read-only Cache <rbd-persistent-read-only-cache>
-   Persistent Write-back Cache <rbd-persistent-write-back-cache>
+   Persistent Write Log <rbd-persistent-write-log>
```
@CongMinYin minor nit: Persistent Write Log is still a form of cache for RBD, to someone not familiar with PWL, I think it would be good to have Cache specified as "Persistent Write Log Cache"

CongMinYin replied:
Make sense, done. `Write back cache` is the earliest name. It was later named `write log` according to the design and implementation. Now, according to our discussion with you, we tend to call it `rbd persistent cache`. The name of `rbd persistent write log cache` is more detailed.

ideepika commented:
https://pulpito.ceph.com/ideepika-2021-12-24_06:25:40-rbd:persistent-writeback-cache-wip-deepika-testing-2021-12-23-2036-distro-default-smithi/

idryomov commented on src/librbd/cache/pwl/AbstractWriteLog.cc:
```diff
@@ -1496,13 +1493,10 @@ bool AbstractWriteLog<I>::check_allocation(
+      ldout(m_image_ctx.cct, 20) << "Waiting for allocation cap (cap="
+                                 << m_bytes_allocated_cap
+                                 << ", allocated=" << m_bytes_allocated
+                                 << ") in write [" << *req << "]" << dendl;
```
Can this result in these messages spamming the log?  It seems that previously `waited_buffers` served as a guard here and this message was printed only once per request.

CongMinYin replied:
verified. Every `process_work()` will output once. Each time `alloc_resource()` failed will exit while(). Log are:
```
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 check_allocation: Waiting for allocation cap (cap=751619276, allocated=751566848) in write [image_extents=[[344457216,262144]], image_extents_summary=[total_bytes=262144, first_image_byte=344457216, last_image_byte=344719360], bl=buffer::list(len=262144,
  buffer::ptr(0~262144 0x55be6d902000 in raw 0x55be6d902000 len 262144 nref 1)
), user_req=0x55be6c230180, m_user_req_completed=0, m_deferred=1, detained=0]
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 wake_up: 
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 process_writeback_dirty_entries: Look for dirty entries
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 can_flush_entry:
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 process_writeback_dirty_entries: Next dirty entry isn't flushable yet
2021-12-29T16:26:40.417+0800 7fd5b5315700 10 librbd::cache::pwl::rwl::WriteLog: 0x55be63758b00 process_work: alloc_fail=1, allocated > high_water=1, allocated_entries > high_water=0
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::rwl::WriteLog: 0x55be63758b00 retire_entries: Look for entries to retire
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::AbstractWriteLog: 0x55be63758b00 can_retire_entry: 
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::rwl::WriteLog: 0x55be63758b00 retire_entries: Nothing to retire
2021-12-29T16:26:40.417+0800 7fd5b5315700 10 librbd::cache::pwl::rwl::WriteLog: 0x55be63758b00 process_work: Retired 0 times
2021-12-29T16:26:40.417+0800 7fd5b5315700 20 librbd::cache::pwl::rwl::WriteLog: 0x55be63758b00 alloc_resources:
```
After my modification https://github.com/ceph/ceph/pull/44157/commits/5573c234b1a001b40a992ba4e4eb017077ea08d7. It may reduce the output to the original similar conditions. Do you think such a change is necessary?

idryomov replied:
If the log isn't spammed then probably not necessary.

CongMinYin replied:
then drop it.

idryomov commented on src/librbd/cache/pwl/rwl/WriteLog.cc:
```diff
@@ -292,8 +290,7 @@ bool WriteLog<I>::initialize_pool(Context *on_finish, pwl::DeferredContexts &lat
       lderr(cct) << "num_small_writes needs to > 2" << dendl;
-      on_finish->complete(-EINVAL);
-      return false;
+      goto error_handle;
```
Give it a meaningful name, something like `err_close_pool`.

CongMinYin commented on src/librbd/cache/pwl/Request.cc:
```diff
@@ -41,10 +41,7 @@ std::ostream &operator<<(std::ostream &os,
-      << "waited_lanes=" << req.waited_lanes << ", "
-      << "waited_entries=" << req.waited_entries << ", "
-      << "waited_buffers=" << req.waited_buffers << "";
+      << "detained=" << req.detained << "";
```
I printed the log to verify. When using `std::ostream` output, we cannot guarantee whether there are spaces before and after the current output statement. So in the whole PWL code, some places have added tail spaces, some places have added head spaces, and some places have forgotten to add. Where do you think it should be added uniformly? Tail? I will make further modifications. For example:
```
2021-12-29T16:10:12.944+0800 7f7651acc700 20 librbd::cache::pwl::Request: 0x56188c62fce0 setup_log_operations: write_req=image_extents=[[990789632,16384]], image_extents_summary=[total_bytes=16384, first_image_byte=990789632, last_image_byte=990806016], bl=buffer::list(len=16384,
	buffer::ptr(0~16384 0x56188e5b4000 in raw 0x56188e5b4000 len 16384 nref 1)
), user_req=0x56188bea1290, m_user_req_completed=0, m_deferred=0, detained=0 m_resources.allocated=1op_set=cell=0x56188d08e7e0, extent_ops_appending=[0x56188c824780, extent_ops_persist=[0x56188bd5f2c0] op_set=0x5618839fa480 operation=0x56188c6f8250

```

here `m_resources.allocated=1op_set=cell=0x56188d08e7e0` lose `" "`.

idryomov replied:
I think head spaces are better.  Presumably you always know where you are starting a new line, in all other places just start with a space.

But my comment wasn't about (potentially missing) spaces -- the line I commented on is appending `""` (i.e. an empty string), not a space.

CongMinYin replied:
I removed it. The original author may want to add a space or `"."`.  Now I add spaces at the head of most `osteam` output, so the tail spaces are no longer needed.  To avoid this form of output: `write_req=image_extents=[[]]`, I referenced other codes and added `[]`. It will show `req=[]`. Now the request ouput shows:

```
2021-12-31T15:50:14.496+0800 7f59f033f700 20 librbd::cache::pwl::Request: 0x5615abb946e0 setup_log_operations: write_req=[image_extents=[64749568,262144], image_extents_summary=[total_bytes=262144, first_image_byte=64749568, last_image_byte=65011712], bl=buffer::list(len=262144,                                                                                                                               
  buffer::ptr(0~262144 0x5615b46de000 in raw 0x5615b46de000 len 262144 nref 1)
), user_req=0x5615abb3e530, m_user_req_completed=0, m_deferred=0, detained=0, waited_lanes=0, waited_entries=0, waited_buffers=0 m_resources.allocated=1 op_set=cell=0x7f59fc98b0e0, extent_ops_appending=[0x5615abb7fb00, extent_ops_persist=[0x5615abb7f800]] op_set=0x5615abb5aa80

```

CongMinYin commented:
@idryomov  updated. There are two places that need your advice. See your previous comments and my replies.

idryomov commented on src/librbd/cache/pwl/ssd/WriteLog.cc:
```diff
@@ -249,7 +249,7 @@ bool WriteLog<I>::initialize_pool(Context *on_finish,
   return true;
 
-error_handle:
+err_close_pool:
```
Nit: in this case a bdev is closed, not a pool.  Normally I try to name these roughly based on the name of the cleanup function or its argument:
```suggestion
err_close_bdev:
```

idryomov commented on doc/rbd/rbd-persistent-write-log-cache.rst:
```diff
@@ -38,11 +40,10 @@ The data is persisted to the cache device when a flush request is received.
 
-To enable the persistent write-back cache, the following Ceph settings
-need to be enabled.::
+To enable the PWL cache, the following Ceph settings need to be enabled.::
```
```suggestion
To enable the PWL cache, set the following configuration settings::
```

idryomov reviewed (commented):
It looks like the new logging in `check_allocation` wouldn't print anything in case the allocation is throttled because of lack of lanes, etc.  I would drop "librbd/cache/pwl: merge output and set output conditions" for now.

CongMinYin commented:
@idryomov add some minor changes in commit  `clean code and update comments`.

- change log `req=` to `req=[]`;
- change error log  to lowercase letter;
- add some head space in osteam ouput;
- remove redundant `[] `when osteam output `image_extents`.

Please check again, thanks.

idryomov commented on src/librbd/cache/pwl/ssd/Request.cc:
```diff
@@ -34,10 +34,10 @@ template <typename T>
-     << "read_bl=" << req.read_bl << ", "
-     << "compare_succeeded=" << req.compare_succeeded << ", "
-     << "mismatch_offset=" << req.mismatch_offset;
+     << " cmp_bl=" << req.cmp_bl << ","
+     << " read_bl=" << req.read_bl << ","
```
In cases like this, that is within a single logging statement, I would merge `","` into the head space.  Less `operator<<` calls, more performance.
```suggestion
     << " cmp_bl=" << req.cmp_bl <<
     << ", read_bl=" << req.read_bl << ","
```

CongMinYin replied:
According to your suggestion, I merged most of the `“<<”` calls that can be merged in PWL in new commit. Please check is it OK?

idryomov reviewed (commented):
Otherwise LGTM!

CongMinYin commented on src/librbd/cache/pwl/LogOperation.cc:
```diff
@@ -20,11 +20,11 @@ GenericLogOperation::GenericLogOperation(utime_t dispatch_time,
+     << "], buf_persist_start_time=[" << buf_persist_start_time
+     << "], buf_persist_comp_time=[" << buf_persist_comp_time
+     << "], log_append_start_time=[" << log_append_start_time
+     << "], log_append_comp_time=[" << log_append_comp_time << "], ";
```
removed it. Add a space at the head of the next output.

idryomov replied:
> Add a space at the head of the next output.

I already commented on two other instances but now I don't see the need for a space before e.g. "(Sync Point) " either.  Could you paste the resulting log messages?

idryomov commented on src/librbd/cache/pwl/LogOperation.cc:
```diff
@@ -269,9 +266,9 @@ WriteLogOperationSet::~WriteLogOperationSet() { }
-  os << "cell=" << (void*)s.cell << ", "
-     << "extent_ops_appending=[" << s.extent_ops_appending << ", "
-     << "extent_ops_persist=[" << s.extent_ops_persist << "]";
+  os << " cell=" << (void*)s.cell
```
Is the space actually needed here?  It looks like it is output as ` " op_set=" << *req.op_set` so I guess not?

idryomov commented on src/librbd/cache/pwl/LogOperation.cc:
```diff
@@ -269,9 +266,9 @@ WriteLogOperationSet::~WriteLogOperationSet() { }
-     << "extent_ops_persist=[" << s.extent_ops_persist << "]";
+  os << " cell=" << (void*)s.cell
+     << ", extent_ops_appending=[" << s.extent_ops_appending
+     << "], extent_ops_persist=[" << s.extent_ops_persist << "]";
```
Did you check the resulting log message?  `extent_ops_appending` and `extent_ops_persist` are `C_Gather *`.  If it is printed as a regular pointer you could just remove the brackets.

idryomov commented on src/librbd/cache/pwl/Request.h:
```diff
@@ -347,9 +347,9 @@ class GuardedRequest {
-    os << "guard_ctx->state=[" << r.guard_ctx->state << "], "
-       << "block_extent.block_start=" << r.block_extent.block_start << ", "
-       << "block_extent.block_start=" << r.block_extent.block_end;
+    os << " guard_ctx->state=[" << r.guard_ctx->state
```
Same here.  It is output as `"req=" << req` and `submitting queued request to blockguard: " << req`, why the extra space?

CongMinYin commented:
@idryomov I have removed three extra head spaces. (What I thought before is that the output may not be in the head in the future. I think too much).  And I removed extra`[]` in `extent_ops_appending `and `extent_ops_persist`. add `[]` for `op_set=[]` The log are:

```
2022-01-05T18:58:47.629+0800 7f6483cb5700 20 librbd::cache::pwl::Request: 0x55ae1b35c840 setup_log_operations: write_req=[image_extents=[431489024,262144], image_extents_summary=[total_bytes=262144, first_image_byte=431489024, last_image_byte=431751168], bl=buffer::list(len=262144,
  buffer::ptr(0~262144 0x55ae1b39e000 in raw 0x55ae1b39e000 len 262144 nref 1)
), user_req=0x55ae126fd120, m_user_req_completed=0, m_deferred=0, detained=0 m_resources.allocated=1 op_set=[cell=0x2072657466612073, extent_ops_appending=0x55ae1273de00, extent_ops_persist=0x55ae1273d8c0]], op_set=0x55ae12718080
```

```
2022-01-05T18:58:47.690+0800 7f64834b4700 20 librbd::cache::pwl::AbstractWriteLog: 0x55ae12856b00 release_guarded_request: submitting queued request to blockguard: guard_ctx->state=[barrier=1, current_barrier=0, detained=0, queued=1], block_extent.block_start=0, block_extent.block_end=18446744073709551615
```

Add another minor change in `src/librbd/BlockGuard.h`, I found the log show:
```
2022-01-05T18:21:01.732+0800 7f623c49c700 20 librbd::cache::pwl::LogMap: 0x56121ada53c8 add_log_entry_locked: block_extent=[block_start = 431489024, block_end = 431751168)
2022-01-05T18:21:01.732+0800 7f623c49c700 20 librbd::cache::pwl::LogMap: 0x56121ada53c8 find_map_entries_locked: block_extent=[block_start = 431489024, block_end = 431751168)
```
So I changed the `)` to `]`. It doesn't become a commit independently. Is is OK?

idryomov commented:
> So I changed the `)` to `]`. It doesn't become a commit independently. Is is OK?

Yes, it's fine.

idryomov commented on src/librbd/BlockGuard.h:
```diff
@@ -33,7 +33,7 @@ struct BlockExtent {
   friend std::ostream& operator<< (std::ostream& os, const BlockExtent& block_extent) {
     os << "[block_start = " << block_extent.block_start << ", "
-       << "block_end = " << block_extent.block_end << ")";
+       << "block_end = " << block_extent.block_end << "]";
```
I'd merge `", "` here as well and drop spaces around `=` to be consistent with most other places.

CongMinYin replied:
Done. And also merged two place `", "`  in this file.

idryomov commented:
https://pulpito.ceph.com/dis-2022-01-06_12:30:28-rbd-wip-dis-testing-distro-default-smithi/
